{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a01e9586-18e2-4ed4-adfb-b3f171cc18a1",
   "metadata": {},
   "source": [
    "Reading big data (~1gb file using bcftools to filter for ERAP2 genomic positions, to remove metadata header and after processing file in Ubuntu, converting to tabular CSV format retaining essential variant information for downstream haplotype assembling, statistical analysis and visualization in Python. \n",
    "**A tab-delimited variant file (ERAP2_mod.csv) was read in 20,000-row chunks using pandas (v2.1.4) in Python3. Chunks were concatenated into a single DataFrame to enable memory-efficient processing.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994eab5f-7827-43d7-b80b-a0695eb7d1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "import seaborn as sns\n",
    "mylist=[]\n",
    "for chunk in pd.read_csv('/hpc/dla_lti/araja/hapsnew/ERAP2homedir/ERAP2_mod.csv', chunksize=20000, sep=\"\\t\"):\n",
    "    mylist.append(chunk)\n",
    "big_dataERAP2=pd.concat(mylist, axis=0)\n",
    "del mylist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11e33251-f554-4618-87f4-1289f8a6c6f3",
   "metadata": {},
   "source": [
    "##**Removing duplicated data if any**##"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "488703b1-fe6e-435a-8b30-82a633cd9d36",
   "metadata": {},
   "outputs": [],
   "source": [
    "big_dataERAP2[big_dataERAP2.duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be7fc33f-d10d-4967-b0f3-5c72caf805ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Removing the columns that are not required\n",
    "big_dataERAP2.pop('QUAL')\n",
    "big_dataERAP2.pop('FILTER')\n",
    "big_dataERAP2.pop('INFO')\n",
    "big_dataERAP2.pop('#CHROM')\n",
    "big_dataERAP2.pop('FORMAT')\n",
    "big_dataERAP2.pop('ID')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6bf5d79-3488-44aa-a9a0-8ebe2f9a1359",
   "metadata": {},
   "source": [
    "**Reading frequency of variants file calculated using PLINK 2.0**\n",
    "plink2 --vcf ERAP2.vcf --freq --out Plink_ERAP2\r",
    "awk '{print $1, $5}' Plink_ERAP2.afreq > Plink_ERAP2.csv\n",
    "\n",
    "\r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dafedb49-73c5-4917-a065-56f616c2ca73",
   "metadata": {},
   "outputs": [],
   "source": [
    "ERAP2_PLINK=pd.read_csv('/hpc/dla_lti/araja/haplotypes/Plink_ERAP2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2080977-3616-4ebd-8b43-3b4d23481130",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Add frequency data to ERAP2 genotype file####\n",
    "#based on POS common column#\n",
    "df1n = big_dataERAP2.set_index('POS')\n",
    "df2n = ERAP2_PLINK.set_index('POS')\n",
    "ERAP2AlleleFreqPlinkMerge=df1n.merge(df2n, how='left', on='POS')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "358cbaa1-af81-4e08-abbd-0f5c09e11c3a",
   "metadata": {},
   "source": [
    "**Downloaded GnomAD file for ERAP2 variants, filtered it for variants of interest and using allele_frequency reported for European population and finally saved in  ERAP2_var_GnomAD** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ea7db81-6a1e-4629-bcba-595a8d282802",
   "metadata": {},
   "outputs": [],
   "source": [
    "ERAP2_gnomeAD=pd.read_csv('/hpc/dla_lti/araja/haplotypes/ERAP2_var_GnomAD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bfe3aaa-8ed0-4731-be70-67a8d8daea0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Adding rsIDs annotation etc by merging GnomAD data####\n",
    "df1 = ERAP2AlleleFreqPlinkMerge\n",
    "df2 = ERAP2_gnomeAD\n",
    "ERAP2_Plink_GnomADmerge = df1.merge(df2, on='POS', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "095edd31-1134-4924-b550-22285f9525b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "####Keeping only Misense variant and splice variant that is just below the exon10 and filtering for MAF >=1/10000####\n",
    "Exclude = ['synonymous_variant','5_prime_UTR_variant', '3_prime_UTR_variant', 'intron_variant', 'stop_gained', 'frameshift_variant']\n",
    "ERAP2_Plink_GnomADmerge1 = ERAP2_Plink_GnomADmerge[~ERAP2_Plink_GnomADmerge.VEP_Annotation.isin(Exclude)]\n",
    "ERAP2_Plink_GnomADmerge2=ERAP2_Plink_GnomADmerge1[ERAP2_Plink_GnomADmerge1.Allele_Freq >= 0.0001]\n",
    "ERAP2_Plink_GnomADmerge2=ERAP2_Plink_GnomADmerge2.dropna(subset=['VEP_Annotation'])\n",
    "ERAP2_Plink_GnomADmerge2.loc[ERAP2_Plink_GnomADmerge2['VEP_Annotation'] == 'missense_variant']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f4dac94-2044-47e6-a371-368c801c867e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###Save the File for further analysis####\n",
    "ERAP2_Plink_GnomADmerge2.to_csv('ERAP2_Full_new_March23_final.csv', sep=\",\", mode='a', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
